contracts/anchor/OriginObserver.sol

```diff
@@ -0,0 +1,82 @@
+pragma solidity >=0.5.0 <0.6.0;
+
+// Copyright 2020 OpenST Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "../anchor/GenesisOriginObserver.sol";
+import "../anchor/StateRootProvider.sol";
+import "../consensus/CoconsensusModule.sol";
+import "../proxies/MasterCopyNonUpgradable.sol";
+
+/**
+ * @title OriginObserver - Contract on the auxiliary chain that stores the
+ *        state root of the origin chain.
+ *
+ * @notice This contract stores the state root and block height from the
+ *         origin chain. This contract is Coconsensus module. Only coconsensus
+ *         contract will be able to anchor the state root.
+ */
+contract OriginObserver is
+    MasterCopyNonUpgradable,
+    GenesisOriginObserver,
+    CoconsensusModule,
+    StateRootProvider
+{
+    /* Constants */
+
+    /** Storage index for genesisBlockNumber */
+    uint8 public constant genesisBlockNumberIndex = uint8(1);
+
+    /** Storage index for genesisStateRoot */
+    uint8 public constant genesisStateRootIndex = uint8(2);
+
+    /** Storage index for MaxStateRootLimitCount */
+    uint8 public constant genesisMaxStateRootLimitCountIndex = uint8(3);
+
+
+    /* External functions */
+
+    /**
+     * @notice Setup function for origin observer contract.
+     *
+     * @dev This can be called by anyone only once. The check in StateRootProvider::setup
+     *      ensures this in its circular buffer setup. Onother check is provided by
```

Comment:
typos
"Another"
"StateRootProvider::anchorStateRoot"

Reply:
perhaps this is in practice not needed, but we briefly touched on this topic more generally:

CoconsensusModule has a constant address for Coconsensus, but we can set it up nonetheless and have a general `mutex` which is required in `onlyConsensus`.

now `anchorStateRoot` would be callable before `setup()` is called

Reply:
but... onlyCoconsensus can call, so it would be correct if Coconsesus has a mutex or general check on the top-level `setup()`

Reply:
@jayeshbairagi I dont believe we have a ticket for `CoconsensusModule` to have a general mutex on being setup, correct?

can you note this for specification discussion after current sprint? then we can now move forward with this approach

Reply:
Fixed typos.